### ISSUE TITLE
History for dmenu

### ISSUE THREAD

**Issue by SicariusNoctis** [DaveDavenport/rofi#747]
**Problem:** History only works for built-in modi such as `drun`. It does not work for piped input in `-dmenu` mode.

**Suggested solution:** Read/save history to a database at a specified path `history-path`. For example:

```
locate | \
    rofi -dmenu -history-path "~/.local/share/rofi/rofi_locate_history"
```

**Example use case:** Higher priority order for commonly opened files.

```
locate \
    | rofi -dmenu -history-path "~/.local/share/rofi/rofi_locate_history" -levenshtein-sort \
    | xargs -0 xdg-open
```

**Rationale:** Currently, history is only enabled for particular modi. This would be a more general way to support custom histories without requiring the user to write a full fledged script to reimplement rofi's history/levenshtein sort features.

This would also solve issues such as #492.

Thanks.

**Comment by DaveDavenport** [comment]
Why should this be in rofi? why not in the script itself?

With dmenu mode rofi should be as 'transparent' as possible, data in -> display it -> result out.

Also adding history as above will only work for very simple 'scripts'.  
For example history is handled for each modi is different, given it has different constraints. (e.g. in drun the  history should also point to a certain application, not desktop file.  It should correctly obey new (higher priority) desktop files and removal. It therefor stores the unique ID of the desktop file.  Run just stores the program name).

You will get the same problem with scripting, maybe you want to only show filename, but need the full path to actually open the file.

**Comment by sardemff7** [comment]
I wanted to add a generic history mechanism for plugins, that could bring some bonuses to the script mode. But the dmenu mode should not be impacted by this.
IOW, it should be possible in the future, but you’ll need a real modi script, not `rofi -dmenu`.

**Comment by liskin** [comment]
@DaveDavenport One reason someone might like to have it in rofi is that there currently doesn't seem to be a way to make Shift+Delete work as seamlessly as it does in rofi natively. I tried to write a script that wraps rofi and uses @YodaEmbedding's https://github.com/YodaEmbedding/frece to add history to it, and I had to rebind Shift+Delete as -kb-custom-1 to make it at least possible to delete entries. This is what I ended up with: https://github.com/liskin/dotfiles/blob/74fed5fca5c2f414a588d2e02880c968eb224615/bin/rofi-dmenu-frece

If I wanted to honor user's preferences I'd additionally need to parse the output of `rofi -dump-xresources`, and I have no idea how to delete entries without exiting rofi, although I suspect it's probably doable using custom modes somehow.

So, um, I'm mostly writing this just to share the solution with others, not to pressure the maintainers to actually implement history for dmenu. Let me know if you have any ideas how I could improve this, and @YodaEmbedding if you want to use this as an example for frece, feel free to do so. :-)
